vpr/src/pack/cluster_legalizer.cpp

```diff
@@ -0,0 +1,1756 @@
+        AtomBlockId root_atom = molecule->atom_block_ids[molecule->root];
+        const std::string& root_atom_name = atom_nlist.block_name(root_atom);
+        if (new_cluster.pb->name != nullptr)
+            free(new_cluster.pb->name);
+        new_cluster.pb->name = vtr::strdup(root_atom_name.c_str());
+        // Move the cluster into the vector of clusters and ids.
+        legalization_cluster_ids.push_back(new_cluster_id);
+        legalization_clusters.push_back(std::move(new_cluster));
+        // Update the molecule to cluster map.
+        molecule_cluster[molecule] = new_cluster_id;
+    } else {
+        // Delete the new_cluster.
+        free_pb(new_cluster.pb);
+        delete new_cluster.pb;
+        free_router_data(new_cluster.router_data);
+        new_cluster_id = LegalizationClusterId::INVALID();
+    }
+
+    return {pack_status, new_cluster_id};
+}
+
+e_block_pack_status ClusterLegalizer::add_mol_to_cluster(t_pack_molecule* molecule,
+                                                         LegalizationClusterId cluster_id) {
+    // Safety asserts to make sure the inputs are valid.
+    VTR_ASSERT_SAFE(cluster_id.is_valid() && (size_t)cluster_id < legalization_clusters.size());
+    VTR_ASSERT(legalization_cluster_ids[cluster_id].is_valid() && "Cannot add to a destroyed cluster");
+    // Ensure that the molecule has not already been placed.
+    VTR_ASSERT(molecule_cluster.find(molecule) == molecule_cluster.end() ||
+               !molecule_cluster[molecule].is_valid());
+    // Safety asserts to ensure that the API was initialized properly.
+    VTR_ASSERT_DEBUG(cluster_placement_stats != nullptr &&
+                     lb_type_rr_graphs != nullptr);
+
+    // Get the cluster.
+    LegalizationCluster& cluster = legalization_clusters[cluster_id];
+    VTR_ASSERT(cluster.router_data != nullptr && "Cannot add molecule to cleaned cluster!");
+    // Set the target_external_pin_util.
+    t_ext_pin_util target_ext_pin_util = target_external_pin_util.get_pin_util(cluster.type->name);
+    // Try to pack the molecule into the cluster.
+    e_block_pack_status pack_status = try_pack_molecule(molecule,
+                                                        cluster,
+                                                        cluster_id,
+                                                        target_ext_pin_util);
+
+    // If the packing was successful, set the molecules' cluster to this one.
+    if (pack_status == e_block_pack_status::BLK_PASSED)
+        molecule_cluster[molecule] = cluster_id;
+
+    return pack_status;
+}
+
+void ClusterLegalizer::destroy_cluster(LegalizationClusterId cluster_id) {
+    // Safety asserts to make sure the inputs are valid.
+    VTR_ASSERT_SAFE(cluster_id.is_valid() && (size_t)cluster_id < legalization_clusters.size());
+    VTR_ASSERT(legalization_cluster_ids[cluster_id].is_valid() && "Cannot destory an already destroyed cluster");
```

Comment:
That is actually common practice for asserts (at least I always did this in LLVM). Strings always resolve to true, so doing an assert like this will actually print a message when the assert is hit; which may make it easier to debug what is going on.